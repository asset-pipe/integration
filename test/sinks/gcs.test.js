'use strict';

const express = require('express');
const { resolve } = require('path');
const supertest = require('supertest');
const Client = require('@asset-pipe/client');
const AssetServer = require('@asset-pipe/server');
const AssetSinkGCS = require('@asset-pipe/sink-gcs');
const { hashArray } = require('../../utils');
const {
    endWorkers,
} = require('../../node_modules/@asset-pipe/server/lib/utils');

async function startTestServer(sink) {
    const app = express();
    const assets = new AssetServer(sink);
    app.use(assets.router());
    return new Promise(resolve => {
        const server = app.listen(() => {
            resolve({
                server,
                port: server.address().port,
            });
        });
    });
}

function closeTestServer(server) {
    return new Promise(resolve => {
        server.once('close', () => resolve());
        server.kill();
    });
}

function clean(data) {
    const reg1 = /http:\/\/127\.0\.0\.1:[0-9]+\//g;
    const reg2 = /"(\/[A-Za-z]+)+\/asset-pipe/gm;
    const str = JSON.stringify(data, null, 2).replace(reg1, '/');
    return JSON.parse(str.replace(reg2, '"asset-pipe'));
}

afterAll(() => endWorkers());

let server;
let port;
let client;
let get;

beforeAll(async () => {
    jest.setTimeout(30000);
    const sink = new AssetSinkGCS(
        {
            projectId: 'asset-pipe-184710', // 679641587805
            keyFilename: './gcs.json',
        },
        'asset-pipe-integration-tests'
    );
    return new Promise(resolve => {
        sink.on('storage info', () => {
            Promise.resolve().then(async () => {
                const started = await startTestServer(sink);
                server = started.server;
                port = started.port;
                get = supertest(`http://127.0.0.1:${port}`).get;
                client = new Client({
                    buildServerUri: `http://127.0.0.1:${port}`,
                });
                resolve();
            });
        });
    });
});

afterAll(async () => {
    await closeTestServer(server);
});

test('javascript upload', async () => {
    expect.assertions(1);
    const upload = await client.uploadFeed([resolve('assets/a.js')]);

    expect(clean(upload)).toMatchSnapshot();
});

test('javascript upload feed fetching', async () => {
    expect.assertions(1);
    const upload = await client.uploadFeed([resolve('assets/a.js')]);
    const { body } = await get(`/feed/${upload.file}`);
    expect(clean(body)).toMatchSnapshot();
});

test('javascript bundling', async () => {
    expect.assertions(1);
    const upload = await client.uploadFeed([resolve('assets/a.js')]);
    const bundle = await client.createRemoteBundle([upload.file], 'js');

    expect(clean(bundle)).toMatchSnapshot();
});

test('javascript bundle fetching', async () => {
    expect.assertions(3);
    const upload = await client.uploadFeed([resolve('assets/a.js')]);
    const bundle = await client.createRemoteBundle([upload.file], 'js');
    const res = await get(`/feed/${bundle.file}`);
    const { headers } = await supertest('https://www.googleapis.com').get(
        `/download/storage/v1/b/asset-pipe-integration-tests/o/${
            bundle.file
        }?alt=media`
    );

    expect(headers['content-type']).toMatch(/application\/javascript/);
    expect(res.headers['content-type']).toMatch(/application\/javascript/);
    expect(clean(res.body)).toMatchSnapshot();
});

test('css upload', async () => {
    expect.assertions(1);
    const feeds = [resolve('assets/style.css')];
    const upload = await client.uploadFeed(feeds);

    expect(clean(upload)).toMatchSnapshot();
});

test('css upload feed fetching', async () => {
    expect.assertions(2);
    const feeds = [resolve('assets/style.css')];
    const upload = await client.uploadFeed(feeds);
    const res = await get(`/feed/${upload.file}`);

    expect(res.headers['content-type']).toMatch(/application\/json/);
    expect(clean(res.body)).toMatchSnapshot();
});

test('css bundling', async () => {
    expect.assertions(1);
    const feeds = [resolve('assets/style.css')];
    const upload = await client.uploadFeed(feeds);
    const bundle = await client.createRemoteBundle([upload.file], 'css');

    expect(clean(bundle)).toMatchSnapshot();
});

test('css bundle fetching', async () => {
    expect.assertions(3);
    const feeds = [resolve('assets/style.css')];
    const upload = await client.uploadFeed(feeds);
    const bundle = await client.createRemoteBundle([upload.file], 'css');
    const res = await get(`/feed/${bundle.file}`).set('Accept', 'text/css');

    const { headers } = await supertest('https://www.googleapis.com').get(
        `/download/storage/v1/b/asset-pipe-integration-tests/o/${
            bundle.file
        }?alt=media`
    );

    expect(headers['content-type']).toMatch(/text\/css/);
    expect(res.headers['content-type']).toMatch(/text\/css/);
    expect(clean(res.body)).toMatchSnapshot();
});

test('bulk javascript bundling', async () => {
    expect.assertions(2);

    const Storage = require('@google-cloud/storage');

    const file1 =
        'b8bde8bab4af3df184ee4cfff61998233c7cd7b0cc9522a982b1925cc11992b5.json';
    const file2 =
        'f4256d8dff32300ea15f41d0ff22cdc71016b2c663b6c7b89ab177df41c38d7f.json';
    const file3 =
        '6171e3331f55b38dd0864fe66d9b2193245ada687b90176ec772fc5d396fbc1c.json';
    const file4 =
        'a8b23f7fafde9106901db75df2c0d28214b6b00c02bb6c383339827bd7819036.json';
    const file5 =
        'def53011b9500e7b9feb1f07ed00bcf68cc6ff04636799b02ed7e4dacfbce9b5.json';

    // Creates a client
    const storage = new Storage({
        projectId: 'asset-pipe-184710',
        keyFilename: './gcs.json',
    });

    const bucket = await storage.bucket('asset-pipe-integration-tests');

    const feed1 = await bucket.file(file1);
    const feed2 = await bucket.file(file2);
    const feed3 = await bucket.file(file3);
    const feed4 = await bucket.file(file4);
    const feed5 = await bucket.file(file5);

    await Promise.all([
        feed1.save(JSON.stringify(require(`../../assets/${file1}`))),
        feed2.save(JSON.stringify(require(`../../assets/${file2}`))),
        feed3.save(JSON.stringify(require(`../../assets/${file3}`))),
        feed4.save(JSON.stringify(require(`../../assets/${file4}`))),
        feed5.save(JSON.stringify(require(`../../assets/${file5}`))),
    ]);

    function doBundle() {
        return client.createRemoteBundle(
            [file1, file2, file3, file4, file5],
            'js'
        );
    }

    let results = [];

    for (let i = 0; i < 2; i++) {
        let a = new Array(10);
        a = a.join(',').split(',');
        const res = await Promise.all(a.map(doBundle));
        results = results.concat(res);
    }

    const uniquefileHashes = new Set(results.map(i => i.file));

    expect(results.length).toBe(20);
    expect(Array.from(uniquefileHashes).length).toBe(1);
});

test('javascript bundling using optimistic bundling', async () => {
    expect.hasAssertions();
    const { resolve } = require;

    const [podletA, podletB, podletC, podletE] = await Promise.all([
        Promise.all([
            client.publishAssets('podletA', [resolve('../../assets/a.js')]),
            client.publishAssets('podletA', [resolve('../../assets/a.css')]),
        ]),
        Promise.all([
            client.publishAssets('podletB', [resolve('../../assets/b.js')]),
            client.publishAssets('podletB', [resolve('../../assets/b.css')]),
        ]),
        Promise.all([
            client.publishAssets('podletC', [resolve('../../assets/c.js')]),
            client.publishAssets('podletC', [resolve('../../assets/c.css')]),
        ]),
        Promise.all([
            client.publishAssets('podletE', [resolve('../../assets/e.js')]),
            client.publishAssets('podletE', [resolve('../../assets/e.css')]),
        ]),
    ]);

    await Promise.all([
        client.publishInstructions('layout', 'js', [
            'podletA',
            'podletB',
            'podletC',
            'podletE',
        ]),
        client.publishInstructions('layout', 'css', [
            'podletA',
            'podletB',
            'podletC',
            'podletE',
        ]),
    ]);

    const jsBundleHash = await hashArray([
        podletA[0].id,
        podletB[0].id,
        podletC[0].id,
        podletE[0].id,
    ]);

    const cssBundleHash = await hashArray([
        podletA[1].id,
        podletB[1].id,
        podletC[1].id,
        podletE[1].id,
    ]);

    const [{ text: jsBundle }, { text: cssBundle }] = await Promise.all([
        get(`/bundle/${jsBundleHash}.js`),
        get(`/bundle/${cssBundleHash}.css`),
    ]);

    expect(jsBundle).toMatchSnapshot();
    expect(cssBundle).toMatchSnapshot();
});
