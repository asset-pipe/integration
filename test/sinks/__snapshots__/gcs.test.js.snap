// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css bundle fetching 1`] = `Object {}`;

exports[`css bundling 1`] = `
Object {
  "file": "8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d.css",
  "uri": "/bundle/8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d.css",
}
`;

exports[`css upload 1`] = `
Object {
  "file": "80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251.json",
  "id": "80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251",
  "uri": "/feed/80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251.json",
}
`;

exports[`css upload feed fetching 1`] = `
Array [
  Object {
    "content": "html, body, div, span, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
abbr, address, cite, code,
del, dfn, em, img, ins, kbd, q, samp,
small, strong, sub, sup, var,
b, i,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section, summary,
time, mark, audio, video {
    margin:0;
    padding:0;
    border:0;
    outline:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

body {
    line-height:1;
}

article,aside,details,figcaption,figure,
footer,header,hgroup,menu,nav,section { 
    display:block;
}

nav ul {
    list-style:none;
}

blockquote, q {
    quotes:none;
}

blockquote:before, blockquote:after,
q:before, q:after {
    content:'';
    content:none;
}

a {
    margin:0;
    padding:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

/* change colours to suit your needs */

ins {
    background-color:#ff9;
    color:#000;
    text-decoration:none;
}

/* change colours to suit your needs */

mark {
    background-color:#ff9;
    color:#000; 
    font-style:italic;
    font-weight:bold;
}

del {
    text-decoration: line-through;
}

abbr[title], dfn[title] {
    border-bottom:1px dotted;
    cursor:help;
}

table {
    border-collapse:collapse;
    border-spacing:0;
}

/* change border colour to suit your needs */

hr {
    display:block;
    height:1px;
    border:0;   
    border-top:1px solid #cccccc;
    margin:1em 0;
    padding:0;
}

input, select {
    vertical-align:middle;
}

h1 {
    color: #000000;
    size: 32px;
}",
    "file": "integration/assets/style.css",
    "id": "8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d",
    "name": "integration",
    "version": "1.0.0",
  },
]
`;

exports[`javascript bundle fetching 1`] = `Object {}`;

exports[`javascript bundling 1`] = `
Object {
  "file": "530d60323a165e72ad9e342778b89cdb5ce8806c41c6f542eb4a1d912016e415.js",
  "uri": "/bundle/530d60323a165e72ad9e342778b89cdb5ce8806c41c6f542eb4a1d912016e415.js",
}
`;

exports[`javascript bundling using optimistic bundling 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(5, 6);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],5:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],6:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(10, 6);

},{\\"./d\\":5}],7:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":8}],8:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}]},{},[1,7,3,6]);
"
`;

exports[`javascript bundling using optimistic bundling 2`] = `
"body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}

body {
    margin: 0;
    padding: 0;
}

.c-title {
    color: green;
}

.e-title {
    color: black;
}"
`;

exports[`javascript upload 1`] = `
Object {
  "file": "149967898754950cb6001298f67fbaa674420d8aa409c48a7ef2c5fb675843d0.json",
  "id": "149967898754950cb6001298f67fbaa674420d8aa409c48a7ef2c5fb675843d0",
  "uri": "/feed/149967898754950cb6001298f67fbaa674420d8aa409c48a7ef2c5fb675843d0.json",
}
`;

exports[`javascript upload feed fetching 1`] = `
Array [
  Object {
    "deps": Object {
      "./d": "320f596434a079912e951a037302087837f241f2896be048e3ccd18d8e6888c9",
    },
    "entry": true,
    "expose": false,
    "file": "asset-pipe/integration/assets/a.js",
    "id": "893576861db8f16e03815f618ca6838cb4af6fd867b10ffa5684e1f1fcd6b467",
    "index": 1,
    "indexDeps": Object {
      "./d": 2,
    },
    "order": 0,
    "source": "'use strict';

const sum = require('./d');

sum(1, 2);
",
  },
  Object {
    "deps": Object {},
    "file": "asset-pipe/integration/assets/d.js",
    "id": "320f596434a079912e951a037302087837f241f2896be048e3ccd18d8e6888c9",
    "index": 2,
    "indexDeps": Object {},
    "source": "'use strict';

module.exports = function sum(a, b) {
    return a + b;
};
",
  },
]
`;
